### ISSUE TITLE
Add a glmmTMB equivalent of lme4::bootMer()

### ISSUE THREAD

**Issue by jhmaindonald** [glmmTMB/glmmTMB#375]
Is this on the "to do" list?

**Comment by bbolker** [comment]
this actually works now, I think, with the *development* (GH) version of `lme4` (which should be going to CRAN sometime next week for CRAN-glitch-fixes anyway) and the [parboot branch of glmmTMB](https://github.com/glmmTMB/glmmTMB/commits/parboot).

If you want to try it out,

```
devtools::install_github("lme4/lme4")
devtools::install_github("parboot@glmmTMB/glmmTMB/glmmTMB")
```

We should presumably (?) also import and re-export `bootMer` so it's available without loading `lme4` ...

**Comment by jhmaindonald** [comment]
That is good news.  I cannot, though, install the parboot branch of glmmTMB:

> devtools::install_github("parboot@glmmTMB/glmmTMB/glmmTMB")
Downloading GitHub repo parboot@glmmTMB/glmmTMB@master
from URL https://api.github.com/repos/parboot@glmmTMB/glmmTMB/zipball/master
Installation failed: Not Found (404)

I tried, also, the following:
> devtools::install_github("glmmTMB/glmmTMB",ref='parboot')
Downloading GitHub repo glmmTMB/glmmTMB@parboot
from URL https://api.github.com/repos/glmmTMB/glmmTMB/zipball/parboot
Installation failed: Does not appear to be an R package (no DESCRIPTION)

Thanks.

**Comment by bbolker** [comment]
Oops, I got the syntax wrong.  Try `"glmmTMB/glmmTMB/glmmTMB@parboot"` ?

**Comment by jhmaindonald** [comment]
Thanks, the installation has now proceeded without difficulty.  In the cases
I have tried with binomial and betabinomial errors, however, the same fixed
effect estimates are returned at each simulation:
```
abc <-
structure(list(Dead = c(95, 88, 99, 99, 91, 112, 119, 120, 120
), Live = c(25, 32, 21, 21, 27, 8, 1, 0, 0), gp = c("A", "A", 
"A", "B", "B", "B", "C", "C", "C")), row.names = c(NA, -9L), class = "data.frame")
ab.BB <- glmmTMB(cbind(Dead,Live)~gp, family=betabinomial(), dispformula=~0+gp,
data=subset(abc, gp!="C"))
ab.Bin <- glmmTMB(cbind(Dead,Live)~gp, family=binomial(), 
                 data=subset(abc, gp!="C"))
bootBin <- bootMer(ab.Bin, function(x)fixef(x)[['cond']], nsim=20)
apply(bootBin$t, 2,range)
##     (Intercept)       gpB
## [1,]    1.285198 0.3998772
## [2,]    1.285198 0.3998772
bootVals <- bootMer(ab.BB, function(x)fixef(x)[['cond']], nsim=20)
apply(bootVals$t, 2,range)
##     (Intercept)       gpB
## [1,]    1.285197 0.4005345
## [2,]    1.285197 0.4005345
```
I tried, also, the example that generated the `fit_zipoisson` object 
(in the 'basic examples' vignette), with results that looked plausible.

**Comment by bbolker** [comment]
OK, the basic diagnosis is clear: this won't work (at present) with a two-column response variable. There are two issues:  (1) I'm  substituting the new response into the data in a really hacky way that uses `deparse()` (I know, @mmaechler , shame on me!) and (2) I need to make sure that two-column responses are handled correctly.

I think a workaround for now would be to compute proportion and total-N variables and use `prop~..., weights=total`, but I agree that this ought to work (and can be made to work).  Ideally I can figure out how to substitute the new response into the `TMBStruc` object returned from `glmmTMB()` when `!doFit`; then `refit.glmmTMB` can just call `fitTMB`.

**Comment by jhmaindonald** [comment]
That does not seem to work, unfortunately:
```
abc$Tot <- with(abc,Dead+Live)
abc$prop <- with(abc,Dead/Tot)
ab.BB <- glmmTMB(prop~gp, weights=Tot, family=betabinomial(), dispformula=~0+gp,
                                 data=subset(abc, gp!="C"))
# Thus far, fine!
coef(summary(ab.BB))
## $cond
##             Estimate Std. Error  z value     Pr(>|z|)
## (Intercept) 1.2851972  0.1288557 9.973928 1.982307e-23
## gpB         0.4005345  0.3395274 1.179682 2.381266e-01
##
## $zi
## NULL
##
## $disp
##    Estimate Std. Error   z value    Pr(>|z|)
## gpA 9.013175  56.457172 0.1596462 0.873159750
## gpB 3.444064   1.052967 3.2708199 0.001072362
##
bootVals <- bootMer(ab.BB, function(x)fixef(x)[['cond']], nsim=20)
## Warning message:
## In bootMer(ab.BB, function(x) fixef(x)[["cond"]], nsim = 20) :
##  some bootstrap runs failed (20/20)
```

**Comment by bbolker** [comment]
OK, I'll see what I can figure out. (At a first look, it seems that there are a *lot* of potential pitfalls in going back and forth between the proportion/weights specification and the successes/failures specification; in particular, `simulate()` still produces two-column output even if the model was specified as proportion/weights ...

**Comment by bbolker** [comment]
resolved in c11d4442fac0bce3a - happy to reopen if bugs/infelicities are discovered
